File: AnyDaf/DafYomiService.swift
import Foundation

/// Fetches today's Daf Yomi from the Sefaria Calendars API
/// and maps it to AnyDaf's canonical tractate names.
@MainActor
struct DafYomiService {

    /// Sefaria tractate name → AnyDaf canonical name (only entries that differ).
    private static let sefariaToAnyDaf: [String: String] = [
        "Eruvin":  "Eiruvin",
        "Taanit":  "Ta\u{2019}anit",
        "Chullin": "Hullin",
        "Middot":  "Middos",
    ]

    struct DafYomi {
        let tractateName: String   // AnyDaf canonical name
        let tractateIndex: Int     // index into allTractates
        let daf: Int
    }

    enum DafYomiError: LocalizedError {
        case networkError(Error)
        case notFound
        case parseError
        case unknownTractate(String)

        var errorDescription: String? {
            switch self {
            case .networkError(let e):       return "Network error: \(e.localizedDescription)"
            case .notFound:                  return "Daf Yomi not found in calendar"
            case .parseError:                return "Could not parse Daf Yomi data"
            case .unknownTractate(let name): return "Unknown tractate: \(name)"
            }
        }
    }

    /// Fetches today's Daf Yomi from Sefaria and returns the matching AnyDaf tractate + daf.
    static func fetchToday() async throws -> DafYomi {
        let url = URL(string: "https://www.sefaria.org/api/calendars")!

        let (data, _): (Data, URLResponse)
        do {
            (data, _) = try await URLSession.shared.data(from: url)
        } catch {
            throw DafYomiError.networkError(error)
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let items = json["calendar_items"] as? [[String: Any]]
        else {
            throw DafYomiError.parseError
        }

        // Find the Daf Yomi entry
        guard let dafYomiItem = items.first(where: { item in
            guard let title = item["title"] as? [String: Any],
                  let en = title["en"] as? String
            else { return false }
            return en == "Daf Yomi"
        }) else {
            throw DafYomiError.notFound
        }

        // Parse tractate + daf from displayValue, e.g. "Menachot 60"
        guard let displayValue = dafYomiItem["displayValue"] as? [String: Any],
              let displayEn = displayValue["en"] as? String
        else {
            throw DafYomiError.parseError
        }

        let parts = displayEn.components(separatedBy: " ")
        guard parts.count >= 2,
              let dafNum = Int(parts.last!)
        else {
            throw DafYomiError.parseError
        }

        let sefariaName = parts.dropLast().joined(separator: " ")
        let anyDafName = sefariaToAnyDaf[sefariaName] ?? sefariaName

        guard let index = allTractates.firstIndex(where: { $0.name == anyDafName }) else {
            throw DafYomiError.unknownTractate(anyDafName)
        }

        return DafYomi(tractateName: anyDafName, tractateIndex: index, daf: dafNum)
    }
}

File: android/app/src/main/java/com/anydaf/data/api/DafYomiService.kt
package com.anydaf.data.api

import com.anydaf.model.allTractates
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject

data class DafYomi(
    val tractateName: String,
    val tractateIndex: Int,
    val daf: Int
)

sealed class DafYomiError : Exception() {
    class NetworkError(cause: Throwable) : DafYomiError()
    object NotFound : DafYomiError()
    object ParseError : DafYomiError()
    class UnknownTractate(val name: String) : DafYomiError()
}

object DafYomiService {

    private val client = OkHttpClient.Builder().build()

    private val sefariaToAnyDaf = mapOf(
        "Eruvin"  to "Eiruvin",
        "Taanit"  to "Ta\u2019anit",
        "Chullin" to "Hullin",
        "Middot"  to "Middos",
    )

    suspend fun fetchToday(): DafYomi = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("https://www.sefaria.org/api/calendars")
            .build()

        val response = try {
            client.newCall(request).execute()
        } catch (e: Exception) {
            throw DafYomiError.NetworkError(e)
        }

        val body = response.body?.string() ?: throw DafYomiError.ParseError

        val json = try { JSONObject(body) } catch (e: Exception) { throw DafYomiError.ParseError }
        val items = json.optJSONArray("calendar_items") ?: throw DafYomiError.ParseError

        // Find the Daf Yomi entry
        var dafYomiItem: JSONObject? = null
        for (i in 0 until items.length()) {
            val item = items.getJSONObject(i)
            val title = item.optJSONObject("title")
            if (title?.optString("en") == "Daf Yomi") {
                dafYomiItem = item
                break
            }
        }
        dafYomiItem ?: throw DafYomiError.NotFound

        val displayValue = dafYomiItem.optJSONObject("displayValue") ?: throw DafYomiError.ParseError
        val displayEn = displayValue.optString("en").ifEmpty { throw DafYomiError.ParseError }

        val parts = displayEn.split(" ")
        if (parts.size < 2) throw DafYomiError.ParseError
        val dafNum = parts.last().toIntOrNull() ?: throw DafYomiError.ParseError

        val sefariaName = parts.dropLast(1).joinToString(" ")
        val anyDafName = sefariaToAnyDaf[sefariaName] ?: sefariaName

        val index = allTractates.indexOfFirst { it.name == anyDafName }
        if (index < 0) throw DafYomiError.UnknownTractate(anyDafName)

        DafYomi(tractateName = anyDafName, tractateIndex = index, daf = dafNum)
    }
}
